### ISSUE TITLE
Redirecting stdout to aiofiles file object produces corrupted data

### ISSUE THREAD

**Issue by uncw** [ronf/asyncssh#594]
Hi @ronf, 

I'm trying to transfer file from remote machine back to my machine. Due to directory permission on the remote machine, I can't use scp or sftp. I use `asyncssh.run` to output the file content to stdout and redirect it to the file object opened using `aiofiles`.

```
async def main():
    async with asyncssh.connect(
        host=host,
        username=user,
        client_keys=['~/.ssh/id_ed25519'],
        known_hosts=None,
    ) as conn:
        async with aiofiles.open('mbfile', 'wb') as f:
            await conn.run(
                'sudo cat /some/directory/mbfile',
                stdout=f,
                stderr=asyncssh.DEVNULL,
                encoding=None,
            )

asyncio.run(main())
```

The `conn.run` completed without error, returncode is 0. The size of the written file matches the remote one's also, but the content is different, part of the file is corrupted. Based on my observation, this only happens to files that are larger than 10MB.

The file contents are different for every single transfer:

```
$ python transfer.py  
$ shasum -a 256 mbfile
2c9fa41d7800c983a084353a504d107891eecbd59bad0d9b1e9f6fafca386302  mbfile
$ python transfer.py  
$ shasum -a 256 mbfile
ba9bab77fa0c25aa7932f411def9a8a1f37ce16b696f9aaed1ff884bea61b551  mbfile
$ python transfer.py  
$ shasum -a 256 mbfile
76178cc65b2107571dd78ecf472b4966cf5563cd66575ca079f2e4d545953492  mbfile
```

I've no clue what's wrong and would greatly appreciate any help.

**Comment by ronf** [comment]
Are the "corrupt" parts all zeroes, or do they seem to contain real data? If it is real data, does it look like maybe the data is written out in the wrong order? I'm not quite seeing how that could happen, but anything you can figure out about the kind of corruption could help us track down the root cause.

If you change from `aiofiles.open()` to a plain `open()`, does that fix the problem?

**Comment by uncw** [comment]
No, they are not all zeroes. It's like chunks of bytes in the wrong order.

And yes, changing to the plain `open()` fixes the problem.

**Comment by ronf** [comment]
I think I see the problem -- the `write()` function in `_AsyncFIleWriter` currently just does a `create_task()` to schedule the write asynchronously. However, if the scheduler ends up running the tasks in a different order than they were created, the data blocks could be written out in the wrong order.

Unfortunately, aiofiles doesn't seem to have a concept of pwrite(), or any other way to atomically do a write at a specific offset within the file. For this to work, I'll probably need to do something like a seek followed by a write while holding an asyncio lock, to make sure that each of the writes ends up in the right place in the final file. Even that could be a problem, though, if a caller were to use aiofiles to write to something like stdout where seeking isn't supported. To properly handle that case, I'd have to using something other than `create_task()` so that I could provide better control of the ordering of the writes.

For now, I think your best bet is to just stick with plain `open()`. While this might block the event loop while the write completes, it's probably not much of an issue unless you are trying to write to a remote network share. I'll see what I can come up with, though.

**Comment by ronf** [comment]
After a bit more thought, I realized I can use an `asyncio.Queue` to solve most of the ordering problem. However, I'm now running into another issue, which is that the process can be closed before all the pending async file operations have completed. If the caller closes the associated file descriptors without waiting for their I/O to complete, the data in the files will end up being incomplete as the remaining writes will be performed on an already closed file. Some kind of `wait_closed` operation is needed here but that's difficult to do since the main APIs to all the redirect-related classes are synchronous right now. That's why in the async case everything had to be run in a separate task. It might be possible to go back through and make the APIs async, but this would be a big change as all the calling functions would have to become asynchronous as well, and at least some of these functions are public APIs where such a change would not be acceptable.

I'll sleep on it and update here if I figure something out.

**Comment by ronf** [comment]
I think I have a fix for this issue, available as commit 8f1fe10 if you'd like to give it a try.

**Comment by uncw** [comment]
It's now working as expected. I initially planned to dig deeper into the internals over the weekend to see if I can come up with something.
Thanks a lot! @ronf.

**Comment by ronf** [comment]
Great - thanks for the confirmation!

There were two issues, really. One was that creating a separate task for each write didn't guarantee that they'd happen in order. However, once a queue was used to make sure order was preserved there was another issue where the aiofile was being closed while there were still writes waiting to occur. Basically, I had to wait in `wait_closed()` on the SSHProcess (called from `run()` in this case) to make sure all async writes were completed before allowing the `wait_closed()` to return.

**Comment by ronf** [comment]
This fix is now available in AsyncSSH 2.14.1.
